src/path.rs

```diff
@@ -0,0 +1,140 @@
+use Renderable;
+use value::Value;
+use context::Context;
+use token::Token;
+use token::Token::*;
+use error::{Error, Result};
+
+#[derive(Debug)]
+pub struct IdentifierPath {
+    value: String,
+    indexes: Vec<Value>,
+}
+
+impl Renderable for IdentifierPath {
+    fn render(&self, context: &mut Context) -> Result<Option<String>> {
+        println!("{:?}", self);
+        let value = context
+            .get_val(&self.value)
+            .ok_or_else(|| Error::Render(format!("{} not found in context", self.value)))?
+            .clone();
+
+        let mut counter = self.indexes.len();
+        let result = self.indexes.iter().fold(Ok(&value), |value, index| {
```

Comment:
1. There a reason you are using `counter` rather than `self.indexes.iter().enumerate()`?
2. [Long term](https://github.com/cobalt-org/liquid-rust/issues/95) I'm wanting to switch it so `Context` to have a trait for the user-provided variables.  Ideally, this would allow lazy lookups.  To get maximize this with indexing, the trait would have some kind of lookup object (slice of Enum(string, index).

Reply:
1. I was affraid to make iterator less readable because of more nesting.
2. `counter` is an internal variable mae to check the depth correctness. It's not used in template and is local to current function. I see no point to place it inside context.

Reply:
2.  Sorry I wasn't clear.  I was referring to the whole handling of value lookups and not counter.  Both comments applied to the exact same line of code, making it harder.

Reply:
I'm sorry, but I can't understand what you mean here. Could you explain a bit more please?

Reply:
`Context.get_val` implements some of the parsing algorithm for looking up a value

1. Ideally we only implement the parsing algorithm in one place
2. I want to copy the minimal part of a value as possible.  Other clients wants to load data from a database (we'd change Context to have a trait to make this possible) and want to load the minimal amount from the database.

So for
```
{{ posts[5].custom.obj.arr[5] }}
```
The current implementation requires copying all of `posts` before indexing into it.

What'd be amazing is if we could do this
```rust
enum PathPart {
    ObjectIndex(&str),
    ArrayIndex(usize),
}

impl Context {
   fn get_val(&self, path: &[PathPart]) -> Option<Value> {
        ...
    }
}
```
This would allow only the leaf to be copied.

Reply:
Another use case for when copying will involve a lot of data: cobalt's `site.data`

Reply:
Oh, I got it. This is a complicated thing actualy. I think we could implement this in current model - where we are not using subexpressions. 
But imagine some perfect case, when we could use subexpressions, like posts[ content[2*3] ]. In this case the subexpression should be rendered with the context as a parameter first. This is a main problem Rust is protecting us from: changing the context using the value from inside the context itself. I'm not sure if introducing a trait could solve it.
I see another kind of a solution here: read-only rendering, i.e. when Context doesn't change. Probably even a separate `RenderMut` should be introduced maybe to do non-readonly rendering where it's required explicitly and which I beleive is a more rare case than immutable one.

Reply:
I feel I'm missing something.  I'm not seeing why `posts[ content[2*3] ]` is a problem, particularly that Rust is protecting us from and what a read-only Context gives us.

My rough run through of how this would work:

For now, this just involves turning the `Identifier` path into `&[PathPart]` and passing that to `Context`.

Long term, when/if we add more complex lookups, we should parse the inner expression first, render it, and make a `PathPart` out of it.

This effectively transforms
```
posts[ content[2*3] ]
```
into
```
{% assign _x = 2*3 %}
{% assign _y = content[_x] %}
posts[_y]
```
(forgive my rough approximation of liquid syntax).

Note: these variables shouldn't actually need to be created in the Context, they should be able to live only in Rust.

and I don't see any problems with doing variable based lookups.  When constructing a `PathPart`, `IdentifierPath` would just do a lookup on the current value.